### PR TITLE
Stale tabs resign the stream mirror instead of fighting newer deploys

### PR DESCRIPTION
**Pull request by jonastemplestein**
## Symptom (live on prd today)

https://os.iterate.com/projects/jeeves/agents/streams/agents/onboarding (and the same page on the `iterate` project) alternates between showing the stream and **"Nothing here yet"**, with the console looping forever:

```
[stream /agents/onboarding browser-feed] local schema version changed; rebuilding member mirror.
stream processor "browser-feed" reduction cache is stale (persisted reducerVersion "", current "0.4.0", ...)
[stream /agents/onboarding browser-stream-mirror] local mirror write failed Error: stream processor progress
  commit fenced: expectedCursorRevision does not match the persisted cursorRevision ...
[stream /agents/onboarding browser-stream-mirror] local mirror ingest failed (attempt 1); resubscribing ...
```

## Root cause: a deploy-boundary livelock between two of your own tabs

The mirror writer lock deliberately embeds a version vector (`stream-leader.ts`), so after a deploy that bumps any member's `schemaVersion` (#2089 bumped browser-feed 4→5, #2121 bumped 5→6 the same day), a still-open pre-deploy tab and a fresh tab hold **different locks over the same shared OPFS database** — both are permanent writers. Then, symmetrically and forever:

1. The fresh tab reads the old tab's recorded schema version → "schema version changed" → `DELETE` the projection (**the "Nothing here yet" flash**) + cursor rewind (which **fences the old tab's in-flight commit**) → records its version.
2. The fenced old tab self-heals, re-elects, reads the *newer* recorded version → also "changed" → deletes + rewinds **downgrade-wards** → records its older version back.
3. Now the fresh tab is fenced. Repeat.

Neither side ever resigns; the old tab never reloads itself. **Erasing prd does not help** — the state lives in browser OPFS and both tabs repopulate it instantly. Closing the stale tab fixes one incident, but it recurs on every schema-bumping deploy.

## Fix: staleness loses deterministically

- At election time, **before touching the database**, the winner queries `navigator.locks.query()` for a **held exclusive** mirror lock for this stream whose version vector **strictly supersedes** its own (every shared member ≥, at least one >). Finding one proves a newer-deploy tab owns the mirror *right now*.
- The stale tab then **resigns**: releases its writer lock and stands by as a follower on the mirror the newer tab maintains (reads keep refreshing via the existing database change channel; appends/runtimeState keep working — same contract as any follower). One console warning suggests reloading.
- A **shared-mode** lock request on the newer tab's own lock name is a zero-polling death watch. Its grant is debounced 2s and re-checked, because a *healthy* newer tab also releases its lock on every routine reconnect — only a writer that *stayed* gone triggers re-election.
- **No evidence of a live newer writer ⇒ behavior unchanged** (rebuild to own schema). That is exactly what heals a prd rollback or a closed newer tab, so rollbacks cannot strand mirrors.

Shared mode on the watch is load-bearing: watches must neither queue behind each other nor read as live writers, otherwise two resigned stale tabs would each mistake the other's watch for a newer writer and neither would ever take over. `findSupersedingMirrorWriterLock` therefore counts held **exclusive** locks only.

The race where both tabs pass the check simultaneously (new tab hasn't started its election yet) still costs one bounded fence/re-elect cycle — the same window the version-vectored lock already accepts by design — and then converges permanently.

## Testing

- New `stream-leader.test.ts` (14 tests): vector build/parse round-trip, the supersedes matrix (including mixed/disjoint vectors answering "not newer" in both directions), and the finder against injected lock snapshots — including the exact #2121 shape (`browser-feed@5` vs `@6`) from both tabs' points of view, and the shared-holder/watch exclusion.
- `pnpm typecheck`, `pnpm lint`, `pnpm format`, full `pnpm test` (200 files / 1960 tests in apps/os) all green.
- Cross-version two-tab behavior is not covered by e2e (previews run one bundle); verified by reasoning through every entry path (fresh election, fence self-heal, resume, rollback, 3-generation tabs) documented in the code comments.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes multi-tab writer election and OPFS mirror coordination at deploy boundaries; mistakes could stall live updates or allow brief dual-writer races, though conservative fallbacks and tests limit blast radius.
> 
> **Overview**
> Fixes a **deploy-boundary livelock** where a pre-deploy tab and a fresh tab each hold a **different version-vectored Web Lock** on the same OPFS mirror and alternately rebuild projections, fence commits, and resubscribe forever.
> 
> After winning its own writer election, a tab now calls **`findSupersedingMirrorWriterLock`** (via `navigator.locks.query()`, **exclusive** holders only) **before** reconcile/DB writes. If another tab holds a lock whose schema vector **strictly supersedes** this bundle, the stale tab **releases writership**, stays connected as a **follower**, and warns to reload.
> 
> **`stream-leader`** adds vector parse/compare (`vectorSupersedes`), the superseding-lock finder, and optional **`shared`** lock mode on `acquireWriterRole` for a **death watch** on the newer writer’s lock. When that watch fires, a **60s** recheck avoids false takeover during the newer tab’s routine reconnect gaps; only if no live newer writer remains does the stale tab **re-elect**.
> 
> If lock query fails or finds nothing, behavior is unchanged (rebuild to own schema). New **`stream-leader.test.ts`** covers the vector matrix and finder edge cases (shared holders ignored, query failures).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 0ffebc5d0e5288a5fd375c4d5ebb2e20a4b00160. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +214 -11 | +127 -6 🟩🟩🟩🟩⬜ |
| Tests | +164 -0 | +148 -0 🟩🟩🟩🟩🟩 |
| **Total** | **+378 -11** | **+275 -6** 🟩🟩🟩🟩🟩 |

<sub>Lines counts every changed line; Significant ignores blank lines and JS comments. Between 0d99ae1 and 1473511, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No preview slot recorded.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `0ffebc5` | [https://auth.iterate-preview-1.com](https://auth.iterate-preview-1.com) | 805.3 KiB | 15.2s | 4.1s |  | 2.5s | [Workflow run](https://github.com/iterate/iterate/actions/runs/205288281740957) | 2026-07-18T17:46:10.149Z | Preview app released. |
| Dummy Petshop | released | `0ffebc5` | [https://dummy-petshop.iterate-preview-1.com](https://dummy-petshop.iterate-preview-1.com) | 198.2 KiB | 5.8s | 5.1s |  | 598ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/205288281740957) | 2026-07-18T17:46:08.276Z | Preview app released. |
| OS | released | `0ffebc5` | [https://os.iterate-preview-1.com](https://os.iterate-preview-1.com) | 4.47 MiB (-0.0 KiB vs main) | 85.7s | 470.0s | 3 retried: Agent scripts can send web-chat messages (with file attachments) and call project tools (vitest x1) · a stateful dynamic worker's alarm fires into its own alarm() method, with native retry (vitest x1) · discarding a new file confirms before permanently removing it (specs x1) | 110.2s | [Workflow run](https://github.com/iterate/iterate/actions/runs/205288281740957) | 2026-07-18T17:46:06.657Z | Preview app released. |
| Streams Example App | released | `0ffebc5` | [https://streams.iterate-preview-1.com](https://streams.iterate-preview-1.com) | 1.49 MiB | 15.1s | 55.6s |  | 9.0s | [Workflow run](https://github.com/iterate/iterate/actions/runs/205288281740957) | 2026-07-18T17:46:16.715Z | Preview app released. |

</details>

### CODE REVIEW

File: apps/os/src/domains/streams/client-libraries/browser/stream-browser-store.ts [hunk 1]
```diff
@@ -1622,12 +1647,82 @@ function createStreamRuntime(
...
+        }).then((stillSuperseding) => {
+          if (disposed || supersededWatch !== watch) return;
+          supersededWatch = undefined;
+          if (stillSuperseding !== undefined) {
+            watchSupersedingWriter(stillSuperseding);
+            return;
+          }
+          scheduleReconnect("superseding mirror writer gone; re-electing", 0);
+        });
+      }, SUPERSEDED_RECHECK_DELAY_MS);
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/streams/client-libraries/browser/stream-browser-store.ts:1714]: Right — a healthy newer writer's release-to-rehold gap can legitimately reach scheduleReconnect's 30s backoff ceiling plus the 15s dial deadline, so 2s read a routine reconnect as death. Fixed in 0ffebc5d0: the recheck now waits 60s, with a comment deriving the bound and the deliberate asymmetry (a false takeover recreates the flicker this PR removes; a slow takeover only delays live updates in an already-stale tab, whose reads keep serving the last-written mirror). A cycling writer that re-holds within the window just re-arms the watch.

File: apps/os/src/domains/streams/client-libraries/browser/stream-browser-store.ts [hunk 2]
```diff
@@ -1622,12 +1647,82 @@ function createStreamRuntime(
...
+  // A newer-deploy tab's writer lock is held: this tab's bundle is the stale
+  // one. Give up writership (keeping the connection — a follower still appends
+  // and reads runtimeState) and let the newer tab own the shared mirror; its
+  // writes reach this tab's reactive queries through the database change
+  // channel like any follower's. Reloading this tab is the real upgrade; until
+  // then, a queued request on the newer tab's own lock is a zero-polling death
+  // watch — granted only when that tab is gone (closed, or navigated away),
+  // which is when re-electing can win cleanly (and, if the newer bundle never
+  // comes back — a rollback — legitimately rebuild to our schema).
+  function resignToSupersedingWriter(newerLockName: string) {
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/streams/client-libraries/browser/stream-browser-store.ts]: Inlined at the election branch in 0ffebc5d0 (watchSupersedingWriter stays a function — it re-arms itself when the recheck finds a still-superseding lock).

[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/streams/client-libraries/browser/stream-browser-store.ts]: Same finding as the earlier thread — resolved by 0ffebc5d0, which inlines the resignation sequence at the election branch.
